Changes usage of _ASSERT for SUCCESS error checking and _ASSERT with and conditional

Gvilla1000-nasa opened the pull request:
<!--- These lines are comments. You can delete or ignore them -->
<!--- NOTE: If your PR is trivial, feel free to delete the "Related Issue" -->
<!---       "Testing" or other sections. -->

<!--- Provide a general summary of your changes in the Title above -->

## Description
Adds new macro for `_ASSERT` when it is used specifically for checks against SUCCESS. The new macro is called `_ASSERT_SUCCESS` and compares two inputs against a success return code (0). 

Additionally, this PR adds a macro for when `_ASSERT ` is used with conditional `.and.` through the use of `_ASSERTS` (which can be renamed to another thing if preferred). 

My original intention with this PR was to create a macro for conditional `.and.` and conditional `.or.` but particularly with `.or.` there are multiple instances with varying number of arguments. I tried experimenting with having optional arguments but was unsuccessful. Because of this, the implementation of the `_ASSERTS` one is limited in usage to only two arguments so perhaps it might be better to just leave the `_ASSERT_SUCCESS` in this PR and remove the other macro?

## Related Issue
<!--- This project primarily accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Usage of a specific _ASSERT case can be converted to a macro.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Tested in intel by changing the error code from good to bad to ensure macro worked as expected. When bad, error message was provided as expected. When good, result was zero diff.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have tested this change with a run of GEOSgcm (if non-trivial)
- [x] I have added one of the required labels (0 diff, 0 diff trivial, 0 diff structural, non 0-diff)
- [x] I have updated the CHANGELOG.md accordingly following the style of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/#how)

tclune reviewed (commented):
- For `_ASSERTS` wouldn't it be clearer to implement it as calling `_ASSERT` with an `.and.`?
- Am thinking `_MASSERT` or `_M_ASSERT` might be better name.
- `_ASSERT_SUCCESS`  seems like could just be `_ASSERTS` - is the intent just to make it a bit shorter on the line? (Not even sure it does that.)

tclune commented on Tests/ExtDataDriverGridComp.F90:
```diff
@@ -545,7 +545,7 @@ subroutine run(this, rc)
 
     call ESMF_GridCompRun(this%gc, userRC=userRC,rc=status)
-    _ASSERT(userRC==ESMF_SUCCESS .and. STATUS==ESMF_SUCCESS,'run failed')
+    _ASSERT_SUCCESS(userRC,STATUS,'run failed')
```
```suggestion
  call ESMF_GridCompRun(this%gc, userRC=userRC,__RC__)
 _VERIFY(userRC)
```

tclune commented on base/MAPL_CubedSphereGridFactory.F90:
```diff
@@ -637,7 +637,7 @@ subroutine check_and_fill_consistency(this, unusable, rc)
            (this%target_lat /= MAPL_UNDEFINED_REAL) .and. &
            (this%stretch_factor /= MAPL_UNDEFINED_REAL) ) then
-         _ASSERT( (this%target_lat >= -90.0) .and. (this%target_lat <= 90), 'latitude out of range')
+         _ASSERTS(this%target_lat >= -90.0, this%target_lat <= 90, 'latitude out of range')
```
```suggestion
         _M_ASSERT(this%target_lat >= -90.0, this%target_lat <= 90.0, 'latitude out of range')
```
```suggestion
         _ASSERT2(this%target_lat >= -90.0, this%target_lat <= 90.0, 'latitude out of range')
```

tclune commented on base/MAPL_Profiler.F90:
```diff
@@ -348,7 +348,7 @@ subroutine MAPL_TimerModeSet(MODE, RC)
       ! Sanity check
       timerMode = mode
-      _ASSERT(timerMode >= MAPL_TimerModeOld .and. timerMode <= MAPL_TimerModeMinMax,'needs informative message')
+      _ASSERTS(timerMode >= MAPL_TimerModeOld, timerMode <= MAPL_TimerModeMinMax,'needs informative message')
```
```suggestion
      _ASSERTS(timerMode >= MAPL_TimerModeOld, timerMode <= MAPL_TimerModeMinMax,'inconsistent value for timerMode')
```

Gvilla1000-nasa commented:
I believe the latest push addresses all the suggestions.

mathomp4 commented:
> I believe the latest push addresses all the suggestions.

@Gvilla1000-nasa I think we'd also discussed making the non-userRC `.and.` asserts into two separate asserts? Though I suppose that could be a separate PR.

Gvilla1000-nasa commented:
> > I believe the latest push addresses all the suggestions.
> 
> @Gvilla1000-nasa I think we'd also discussed making the non-userRC `.and.` asserts into two separate asserts? Though I suppose that could be a separate PR.

Working on it right now. Found a missing .and. with error checking as well

tclune commented on base/MAPL_sun_uc.F90:
```diff
@@ -2994,7 +2994,8 @@ subroutine MAPL_SunGetDaylightDurationMax(ORBIT,LATS,DAYL,currTime,CLOCK,RC)
          OBQ = ORBIT%OB * (MAPL_PI/180.)
       endif
-      _ASSERT(0. <= OBQ .and. OBQ < MAPL_PI, 'Strange obliquity!')
+      _ASSERT(OBQ >= 0, 'Obliquity less than 0 detected!')
```
Not going to ask for changes here, but we really need to develop a consistent policy about how to word such messages.  And then _gradually_ propagate.
